Mdx stanza

NathanReb opened the pull request:
This is still draft as I need to either mock mdx for the tests or disable them by default.

I also need to get some stuff merged on the mdx side and cut a release before we can merge this.

Finally some mdx users call dune within mdx files and I want to make sure this won't break too much stuff either.

NathanReb committed:
Add basic mdx stanza

Signed-off-by: Nathan Rebours <nathan.p.rebours@gmail.com>

NathanReb committed:
Allow mdx tests to depend on local packages

Signed-off-by: Nathan Rebours <nathan.p.rebours@gmail.com>

NathanReb committed:
Use ocaml-mdx deps to get dependencies of mdx files

Signed-off-by: Nathan Rebours <nathan.p.rebours@gmail.com>

NathanReb commented:
Note to myself: I need to bump the dune lang as well now that 2.2.0 has been released.

aalekseyev commented:
Is having mdx enabled in CI problematic for some reason? (is it particularly big or broken?)
If adding it to opam_install_test_deps .travis-ci.sh is sufficient to make tests pass, I think that's probably the way to go instead of disabling tests.

NathanReb commented:
> Is having mdx enabled in CI problematic for some reason?
It shouldn't be, you're right. I need to release the version I used to work on the stanza but then I'll add mdx to the CI.

NathanReb committed:
Add changelog entry for mdx stanza

Signed-off-by: Nathan Rebours <nathan.p.rebours@gmail.com>

rgrinberg commented on src/dune/mdx.ml:
```diff
@@ -0,0 +1,211 @@
+         separated items"
+
+  let read files =
+    let open Files in
```
Let's get rid of this open and just annotate the type of `files : Files.t`

rgrinberg commented on src/dune/mdx.ml:
```diff
@@ -0,0 +1,211 @@
+(** Generates the rules for a single [src] file covered covered by the given
+    [stanza]. *)
+let gen_rules_for_single_file ~sctx ~dir ~mdx_prog ~stanza src =
+  let open Files in
```
Same as above. The open is confusing

rgrinberg reviewed (commented):
Looks good. Just some minor changes required and let's make sure the manual is up to date.

NathanReb commented on src/dune/mdx.ml:
```diff
@@ -0,0 +1,211 @@
+
+  let dir_without_files_dep =
+    let pred =
+      Predicate.create ~id:(lazy (String "false")) ~f:(fun _ -> false)
```
Good idea!

NathanReb committed:
Add support for mdx prelude option

Signed-off-by: Nathan Rebours <nathan.p.rebours@gmail.com>

rgrinberg commented on src/dune/mdx.ml:
```diff
@@ -0,0 +1,247 @@
+       field "files" Predicate_lang.Glob.decode ~default:default_files
+     and+ packages = field_o "packages" (repeat Package.Name.decode)
+     and+ preludes = field_o "preludes" (repeat Prelude.decode) in
+     let packages = Option.value ~default:[] packages in
```
I believe you can use `field ~default:[]` to simplify the definition of packages and preludes.

rgrinberg commented on src/dune/mdx.ml:
```diff
@@ -0,0 +1,247 @@
+  let diff_action =
+    let+ () = Build.path (Path.build files.src)
+    and+ () = Build.path (Path.build files.corrected) in
+    Action.diff ~optional:false (Path.build files.src)
```
you might as well do:

```
let src = Path.build files.src
and corrected = Path.build files.corrected
in
```

NathanReb committed:
Bump dune lang version to 2.4

Signed-off-by: Nathan Rebours <nathan.p.rebours@gmail.com>

NathanReb committed:
Upgrade mdx deps parser to use sexp instead of adhoc format

Signed-off-by: Nathan Rebours <nathan.p.rebours@gmail.com>

NathanReb commented:
I just changed the communication format to use canonical sexp instead of something adhoc and fragile.

If that seems fine by you I'll cut an MDX release so that I can fix the CI here and we can finally merge!

NathanReb committed:
Use field ~default instead of field_o

Signed-off-by: Nathan Rebours <nathan.p.rebours@gmail.com>

NathanReb committed:
Simplify mdx diff action definition

Signed-off-by: Nathan Rebours <nathan.p.rebours@gmail.com>

NathanReb commented:
Still have to write the documentation though!

NathanReb committed:
Drop mli files from documentation and default files in mdx stanza

I thought the mli support would make it to 1.6.0 but it did not.
I'm doing in this in a separate commit so it can be reverted later.

Signed-off-by: Nathan Rebours <nathan.p.rebours@gmail.com>

NathanReb committed:
Add documentation for the mdx stanza

Signed-off-by: Nathan Rebours <nathan.p.rebours@gmail.com>

NathanReb commented:
I added some documentation, removed mli files from the doc and the default files as the mli support didn't make it to `mdx.1.6.0`.

Last step is for opam-repo to properly update so that we can install mdx.1.6.0 in the CI to get the test to pass there!

NathanReb commented:
I'm undrafting as it is essentially ready for review!

rgrinberg commented on src/dune/mdx.ml:
```diff
@@ -0,0 +1,236 @@
+
+  let dir_without_files_dep =
+    let pred =
+      Predicate.create ~id:(lazy (String "false")) ~f:(fun _ -> false)
```
This should be moved to `Predicate.false_` (like `Predicate.true_)

rgrinberg committed:
_

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

NathanReb commented:
`mdx.1.6.0` has been released so the CI should go green now!

NathanReb commented:
Just learned about the `runtest-no-deps` thing. Now it should be good!

NathanReb commented:
And there we go! @rgrinberg can I go ahead and merge?

rgrinberg commented:
Yup
On Feb 21, 2020, 11:01 AM +0000, Nathan Rebours <notifications@github.com>, wrote:
> And there we go! @rgrinberg can I go ahead and merge?
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
